enhancements/insights/insights-operator-runtime.md

```diff
@@ -0,0 +1,532 @@
+
+* As a product manager in the AI space, I can measure the application platforms/frameworks/languages used in clusters or namespaces running OpenShift AI to understand what users are coding applications in.
+
+* As a middleware product manager, I can track the number of applications built using Quarkus to understand our return on investment from an OpenShift workload perspective.
+
+* As a RHEL product manager, I can follow trends in the use of other Linux containers to inform my product roadmap.
+
+* As an OpenShift product marketing manager, I want to be able to produce content that shows users use OpenShift as a  platform for all types of workloads.
+    
+
+### Goals
+
+The goal of this proposal is to inspect running containers and pop a few key pieces of data to identify the workload that they are running. This data will enrich and complement the information already gathered by the Insights Operator to get a more accurate representation of workloads running on OpenShift.
+
+### Non-Goals
+
+* It is not a goal of this proposal to scan container for security purposes
+* It is not a goal of this proposal to get runtime information specific to a given customer or application
+* It is not a goal of this proposal to capture personal sensitive information
+
+## Proposal
+
+This enhancement proposal introduces additional data to identify the technology stack of running containers.
+
+The existing `workloadContainerShape` type that contains workload information about a container will be enhanced with a new `workloadRuntimeInfoContainer` field:
+
+```go
+// workloadContainerShape describes the shape of a container which includes
+// a subset of the data in the container.
+// TODO: this may desirable to make more precise with a whole container hash
+//
+//  that includes more of the workload, but that would only be necessary if
+//  it assisted reconstruction of type of workloads.
+type workloadContainerShape struct {
+   // ImageID is the content addressable hash of the image as observed from
+   // the status or the spec tag.
+   ImageID string `json:"imageID"`
+   // FirstCommand is a hash of the first value in the command array, if
+   // any was set.
+   FirstCommand string `json:"firstCommand,omitempty"`
+   // FirstArg is a hash of the first value in the arguments array, if any
+   // was set.
+   FirstArg string `json:"firstArg,omitempty"`
+   // Runtime info on the container
+   RuntimeInfo workloadRuntimeInfoContainer `json:"runtimeInfo,omitempty"`
+}
+```
+
+This new `workloadRuntimeInfoContainer` is defined as:
+
+```go
+type workloadRuntimeInfoContainer struct {
+	// Hash of the identifier of the Operating System (based on /etc/os-release ID)
+	Os string `json:"os,omitempty"`
+	// Hash of the version identifier of the Operating System (based on /etc/os-release VERSION_ID)
```

Comment:
`/etc/os-release` is [a systemd-ism][1].  Maybe all the OS/version you care about use systemd, or have adopted `/etc/os-release` independently of systemd?  But we also support [Windows containers][2], and it may be worth adding a section under *Risks and Mitigations* that goes into how you expect to handle workloads outside of the set that you're able to successfully interpret.

[1]: https://www.freedesktop.org/software/systemd/man/latest/os-release.html
[2]: https://docs.openshift.com/container-platform/4.15/windows_containers/index.html

Reply:
That's a good point, I'll update the RFE to point out this limitations. Even something that looks as simple as OS versions can not be assumed to be reliably collected. There are images build from `scratch`, non-systemd Linux distributions and as you mention Windows containers.
As we are not able to collect OS data from `/etc/os-release` for these types of containers, we will still have "unknown OS" containers but we will at least have an idea of the size of that slice compared to the overall workloads running on OpenShift.
However your point about Windows containers makes me realise that the daemonset should only be deployed on Linux worker nodes

Reply:
I amended the RFE and added a section in the "Variation and form factor considerations":

> This enhancement only gather runtime info from Linux worker nodes and will not touch Windows worker nodes. 
> The DaemonSet will be configured to only be deployed on Linux worker nodes (having the label `beta.kubernetes.io/os=linux`).